File: base44/shared/indicators.ts
// Real technical indicator formulas — ported verbatim from src/lib/indicators.js.
// Computed from actual OHLCV price series. No LLM, no estimation.

export function sma(values, period) {
  if (!values || values.length < period) return null;
  let sum = 0;
  for (let i = values.length - period; i < values.length; i++) sum += values[i];
  return sum / period;
}

export function emaSeries(values, period) {
  if (!values || !values.length) return [];
  const k = 2 / (period + 1);
  const out = [values[0]];
  for (let i = 1; i < values.length; i++) {
    out.push(values[i] * k + out[i - 1] * (1 - k));
  }
  return out;
}

export function ema(values, period) {
  const s = emaSeries(values, period);
  return s.length ? s[s.length - 1] : null;
}

export function rsi(closes, period = 14) {
  if (!closes || closes.length < period + 1) return null;
  let gains = 0;
  let losses = 0;
  for (let i = 1; i <= period; i++) {
    const diff = closes[i] - closes[i - 1];
    if (diff >= 0) gains += diff;
    else losses -= diff;
  }
  let avgGain = gains / period;
  let avgLoss = losses / period;
  for (let i = period + 1; i < closes.length; i++) {
    const diff = closes[i] - closes[i - 1];
    const gain = diff > 0 ? diff : 0;
    const loss = diff < 0 ? -diff : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
  }
  if (avgLoss === 0) return 100;
  const rs = avgGain / avgLoss;
  return 100 - 100 / (1 + rs);
}

export function macd(closes, fast = 12, slow = 26, signal = 9) {
  if (!closes || closes.length < slow + signal) return null;
  const fastEma = emaSeries(closes, fast);
  const slowEma = emaSeries(closes, slow);
  const macdLine = fastEma.map((v, i) => v - slowEma[i]);
  const signalLine = emaSeries(macdLine, signal);
  const i = macdLine.length - 1;
  return {
    macd: macdLine[i],
    signal: signalLine[i],
    histogram: macdLine[i] - signalLine[i],
  };
}

export function bollinger(closes, period = 20, mult = 2) {
  if (!closes || closes.length < period) return null;
  let sum = 0;
  for (let i = closes.length - period; i < closes.length; i++) sum += closes[i];
  const mean = sum / period;
  let variance = 0;
  for (let i = closes.length - period; i < closes.length; i++) variance += (closes[i] - mean) ** 2;
  variance /= period;
  const sd = Math.sqrt(variance);
  const last = closes[closes.length - 1];
  const width = mult * sd;
  return {
    upper: mean + width,
    middle: mean,
    lower: mean - width,
    percentB: width > 0 ? ((last - (mean - width)) / (2 * width)) * 100 : 50,
  };
}

export function atr(highs, lows, closes, period = 14) {
  if (!closes || closes.length < period + 1) return null;
  const trs = [];
  for (let i = 1; i < closes.length; i++) {
    const tr = Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1])
    );
    trs.push(tr);
  }
  let a = 0;
  for (let i = 0; i < period; i++) a += trs[i];
  a /= period;
  for (let i = period; i < trs.length; i++) {
    a = (a * (period - 1) + trs[i]) / period;
  }
  return a;
}

export function momentum(closes, period = 14) {
  if (!closes || closes.length < period + 1) return null;
  const past = closes[closes.length - 1 - period];
  return ((closes[closes.length - 1] - past) / past) * 100;
}

export function volatility(closes, period = 20) {
  if (!closes || closes.length < period + 1) return null;
  const returns = [];
  for (let i = closes.length - period; i < closes.length; i++) {
    returns.push((closes[i] - closes[i - 1]) / closes[i - 1]);
  }
  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((a, b) => a + (b - mean) ** 2, 0) / returns.length;
  return Math.sqrt(variance) * Math.sqrt(365) * 100;
}

File: tradepulse/strategy/indicators.py
"""Real technical indicator formulas -- direct port of base44/shared/indicators.ts.
Computed from actual OHLCV price series. No LLM, no estimation.

Operates on plain floats (matching the TS/IEEE-754 numeric semantics being
ported) rather than Decimal -- these are dimensionless statistical scores,
not money, so float precision is the correct and faithful choice here.
"""

from __future__ import annotations

from dataclasses import dataclass


def sma(values: list[float], period: int) -> float | None:
    if not values or len(values) < period:
        return None
    return sum(values[-period:]) / period


def ema_series(values: list[float], period: int) -> list[float]:
    if not values:
        return []
    k = 2 / (period + 1)
    out = [values[0]]
    for value in values[1:]:
        out.append(value * k + out[-1] * (1 - k))
    return out


def ema(values: list[float], period: int) -> float | None:
    series = ema_series(values, period)
    return series[-1] if series else None


def rsi(closes: list[float], period: int = 14) -> float | None:
    if not closes or len(closes) < period + 1:
        return None
    gains = 0.0
    losses = 0.0
    for i in range(1, period + 1):
        diff = closes[i] - closes[i - 1]
        if diff >= 0:
            gains += diff
        else:
            losses -= diff
    avg_gain = gains / period
    avg_loss = losses / period
    for i in range(period + 1, len(closes)):
        diff = closes[i] - closes[i - 1]
        gain = diff if diff > 0 else 0.0
        loss = -diff if diff < 0 else 0.0
        avg_gain = (avg_gain * (period - 1) + gain) / period
        avg_loss = (avg_loss * (period - 1) + loss) / period
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100 - 100 / (1 + rs)


@dataclass(frozen=True, slots=True)
class MacdResult:
    macd: float
    signal: float
    histogram: float


def macd(closes: list[float], fast: int = 12, slow: int = 26, signal: int = 9) -> MacdResult | None:
    if not closes or len(closes) < slow + signal:
        return None
    fast_ema = ema_series(closes, fast)
    slow_ema = ema_series(closes, slow)
    macd_line = [f - s for f, s in zip(fast_ema, slow_ema)]
    signal_line = ema_series(macd_line, signal)
    i = len(macd_line) - 1
    return MacdResult(macd=macd_line[i], signal=signal_line[i], histogram=macd_line[i] - signal_line[i])


@dataclass(frozen=True, slots=True)
class BollingerResult:
    upper: float
    middle: float
    lower: float
    percent_b: float


def bollinger(closes: list[float], period: int = 20, mult: float = 2) -> BollingerResult | None:
    if not closes or len(closes) < period:
        return None
    window = closes[-period:]
    mean = sum(window) / period
    variance = sum((c - mean) ** 2 for c in window) / period
    sd = variance**0.5
    last = closes[-1]
    width = mult * sd
    percent_b = ((last - (mean - width)) / (2 * width)) * 100 if width > 0 else 50.0
    return BollingerResult(upper=mean + width, middle=mean, lower=mean - width, percent_b=percent_b)


def atr(highs: list[float], lows: list[float], closes: list[float], period: int = 14) -> float | None:
    if not closes or len(closes) < period + 1:
        return None
    trs = []
    for i in range(1, len(closes)):
        tr = max(highs[i] - lows[i], abs(highs[i] - closes[i - 1]), abs(lows[i] - closes[i - 1]))
        trs.append(tr)
    value = sum(trs[:period]) / period
    for i in range(period, len(trs)):
        value = (value * (period - 1) + trs[i]) / period
    return value


def momentum(closes: list[float], period: int = 14) -> float | None:
    if not closes or len(closes) < period + 1:
        return None
    past = closes[-1 - period]
    if past == 0:
        return None
    return ((closes[-1] - past) / past) * 100


def volatility(closes: list[float], period: int = 20) -> float | None:
    if not closes or len(closes) < period + 1:
        return None
    n = len(closes)
    returns = [(closes[i] - closes[i - 1]) / closes[i - 1] for i in range(n - period, n)]
    mean = sum(returns) / len(returns)
    variance = sum((r - mean) ** 2 for r in returns) / len(returns)
    return (variance**0.5) * (365**0.5) * 100
